Bonus: refactored DepositPeriod and ContestationPeriod

ch1bo opened the pull request:
Stacked diff on top of #1974

Refactorings I encountered after fixing the actual issue at hand and introducing `--deposit-period`.

---

* [x] CHANGELOG update not needed
* [x] Documentation update not needed
* [x] Haddocks updated
* [x] No new TODOs introduced

ch1bo committed:
Consolidate developer docs on incremental commits

We were documenting off-/on-chain interaction of these things a bit all
over the place. This makes the protocol.md description more central to
considerations around the design of incremental commit/decommit protocol
extensions.

ch1bo committed:
Write rollback resistence section

ch1bo committed:
Re-use propIsDistributive

ch1bo committed:
Improve counterexample for incorret balance error

The previous code was summing up values which makes it hard to spot
inconsistencies on the actual UTxO set.

ch1bo committed:
Add a Deposit action and simulateDeposit

This will allow us to express clearly the initial step of an incremental
commit. The deposit is not tied to a specific party and thus a deposit
may be done using any available utxo to deposit. The HeadId is tracked
in the model as a symbolic variable.

ch1bo committed:
Delete unused showFromAction

ch1bo committed:
Make waitUntilMatch use a Maybe predicate

This is more powerful and can yield a consistent result observed on all
test clients.

ch1bo committed:
Require headId in simulateCommit

Not sure if this an improvement, but querying the head state feels odd.
Also, this makes simulateCommit less stateful, which further indicates
that draftCommitTx (et al) should actually be possible without this
special provision on the simulated chain.

ch1bo committed:
Improve counterexample of invalid txs in MockChain

ch1bo committed:
Discard invalid deposits in head logic

ch1bo committed:
Change CommitRecorded fields

The update of localUTxO was particularly unsettling.

ch1bo committed:
Move snapshotting of deposits to Tick handler

This should be fairly similar to what we had before, with the difference
snapshots would only be requested after the next Tick (block) seen on
chain.

While this might seem cumbersome now, this is actually what a hydra-node
needs to do and wait even longer before a deposit may be accepted.

ch1bo committed:
Wait for full incremental commit finalization in Model

ch1bo committed:
Use previous snapshot utxoToCommit if not cleared

ch1bo committed:
Improve error message of waitUntilMatch

ch1bo committed:
Bring back a single pending utxo to commit (Uα)

This was the previous 'pendingDeposits', but i was mislead by that being
a map. Now we have two separate things in our state: pendingDeposits to
track deposits we see on chain and the currentDepositUTxO that is used
for incremental commits.

ch1bo committed:
Improve ModelSpec counterexample of HUnitFailures

ch1bo committed:
Let waitUntilMatch fail again on client message

This is particularly useful to see PostTxOnChainFailed messages which
are of this kind.

ch1bo committed:
Ignore empty deposits in HeadLogic

This makes the honest node not even try to snapshot such deposits.
However, we still need to prevent signing of such snapshots.

ch1bo committed:
Introduce a Deposit and DepositStatus type

ch1bo committed:
Only request snapshot for active deposits

This is hiding the problem that other non-honest nodes could also
request such snapshots!

ch1bo committed:
Introduce a DepositActivated and DepositExpired state changes

These are not strictly needed for the logic, but to update the
pendingDeposits in the state. Also they are very helpful to debug
failing tests.

Only the DepositExpired is mapped to a server output becaues activation
is more of an internal process.

The pattern of determining newActive/newExpired and wanting to determine
the next active given the already updated state is maybe a hint to a
more monadic way of writing the logic functions?

ch1bo committed:
Use simulateCommit in BehaviorSpec

This is needed such that both nodes see the deposit because now only the
second snapshot will be finalizing the commit.

ch1bo committed:
Return a txid from simulateDeposit and fix all behavior tests

Also indicated tests where there is no real need of multiple nodes (is
it needed for any of these?).

ch1bo committed:
Add Show and IsString intances to Hydra keys

This allows us to easily create unit tests from model test failures. See
module documentation for details.

ch1bo committed:
Tick logic handles now directly expired deposits

ch1bo committed:
Refactor ModelSpec to have less functions

Following the example of quickcheck-dynamic:
https://github.com/input-output-hk/quickcheck-dynamic/blob/main/quickcheck-dynamic/test/Spec/DynamicLogic/RegistryModel.hs
with a re-usable propDL function.

ch1bo committed:
Fix HasVariables instances

This makes the counterexample correctly bind symbolic variables.

ch1bo committed:
Leave notes about tests to write

ch1bo committed:
Write a mutation test for deposit observation

Any change to the deposit output value without also updating the
recorded commit in the datum should invalidate the deposit observation.

ch1bo committed:
Check deposit value on observation

This makes sure that the deposited value as claimed in the datum is
actually available at the deposit script output to be incremented.

ch1bo committed:
Update a few haddocks on utxo generators

ch1bo committed:
Generate healthy deposit transactions

This makes sure we test a wider range of transactions and we can easily
switch to this approach for deposits as the mutations are fully defined
on the mutated tx/utxo.

ch1bo committed:
Implement behavior test for deadlines in the past

ch1bo committed:
Implement behavior test for deadlines too early

ch1bo committed:
Move Environment type to hydra-node

This type is only used by hydra-node and consequently the hydra-tx
package should not include it.

ch1bo committed:
Update assertions of deadline too soon case

ch1bo committed:
Expire deposits one contestation period before deadline

This also uses the default contestation period of one hour in behavior
tests, which we can do because time is simulated a lot faster.

ch1bo committed:
Leave a note on a changed decommit test assertion

ch1bo committed:
Define genUTxO in terms of genTxOut and bound ada

The ada in values generated by cardano-ledger is often using maxBound of
Word64 and this goes quickly out of bounds when summed up. This changes
genTxOut to only produce realistic ada values (<< max supply).

ch1bo committed:
Fix mutation tests

ch1bo committed:
Fix lints

ch1bo committed:
Define capped realistic ada using size and max supply

Also ensures the minimum lovelace in an output using default pparams.

ch1bo committed:
Write changelog and decide what is still todo

ch1bo committed:
Change Deposit in Model tests to always correct deadline

We have a behavior test that deals with deadlines in the past now. If we
generate deadlines in the model, we would need to evolve the model state
depending on whether we expect it to be expired or not. Complexity that
is currently not needed.

ch1bo committed:
Add a chainTime to DepositExpired and fix Model

This helps debugging. The model was picking a contestation period lower
than the mock chain block time, which is problematic. We should
frontload the decision about a dedicated deposit period into SeedWorld
where we would know what kind of network we are on.

ch1bo committed:
Delete a head logic test which does not apply anymore

The decision whether to snapshot a deposit (= do a commit) or not is not
upon observing a deposit anymore.

ch1bo committed:
Rename Commit{Recorded,Recovered} state events to Deposit{Recorded,Recovered}

This is more consistent to what is happening in the system. This keeps
the server outputs unchange but adds notes on where inconsistencies are.
Those should be addressed by also updating HTTP endpoints.

ch1bo committed:
Fix api.yaml

ch1bo committed:
Resolve weeds

ch1bo committed:
Write a test for preventing too soon snapshots approval

This scenario simulates an adverserial or differently configured node to
request a snapshot to increment a deposit only they deem fine.

ch1bo committed:
Fix deposits going active and expired all the time

This was due to non mutually exclusive states predicates.

ch1bo committed:
Fix time not advancing in open state

The fact that we did forget this hints that we should think about
composing event handlers differently. For example: compose the general
tick handler with state-specific tick handlers

ch1bo committed:
Refactor deposit e2e tests a bit

ch1bo committed:
Fix approving of snapshot with deposits with deadline too soon

This is changing the ReqSn message sent between hydra nodes to only
include a deposit transaction id to be considered. This makes it natural
to only approve snapshots where the deposit txid is known and active.

This way we also handle the problem of requesting snapshots with empty
deposit utxo by design.

ch1bo committed:
Fix schemas and golden files due to changed ReqSn message

This also moves to use ADTArbitrary as the simple golden file was not
consistently incuding the ReqSn mesage type.

ch1bo committed:
Rename --deposit-deadline to --deposit-period

This parameter is used to set the deadline, but also to check whether
deposits are still active.

ch1bo committed:
Use a constant deposit period in Hydra.Model

We are not testing deposits in particular there.

ch1bo committed:
Add explanations for --deposit-period

ch1bo committed:
Move DepositPeriod to hydra-node package

It's only used there, so no point in keeping it in hydra-tx

ch1bo committed:
Remove depositPeriod from withHydraCluster

ch1bo committed:
Remove depositPeriod from chainConfigFor

This is now using the default period unless overridden.

ch1bo committed:
Simplify usage of contestation period

By adding Enum, Num and Integral instances we can easily create, compute
and convert contestation periods. The same pitfalls as with Natural do
exist.

This also moves the specs to the hydra-tx package (where the data type is).

ch1bo committed:
Small fix on Deposit modeling of Hydra.Model

noonio commented on hydra-tx/src/Hydra/Tx/ContestationPeriod.hs:
```diff
@@ -9,14 +9,35 @@ import Hydra.Data.ContestationPeriod qualified as OnChain
+
+  fromEnum (UnsafeContestationPeriod n) = fromEnum n
+
+instance Num ContestationPeriod where
```
Any reason not to use something like this - https://hackage.haskell.org/package/integer-types-0.1.4.1/docs/Integer-Positive.html ?

ch1bo replied:
That package looks good! I did not know about it

ch1bo replied:
No energy left to do it on this PR. Feel free ro refactor more.

ch1bo committed:
Seed and consume UTxO deposit in model tests

We must not add decommitted UTxO to this as the 'Payment'
transaction model may result in conflicting transaction ids / TxIns when
running the model in the real world.

ch1bo committed:
Make deposit UTxO smaller in increment mutation tests

generateWith uses size 30 and this is too big of utxo to deposit.

ch1bo committed:
Address comments on documentation
